app/pidfile/pidfile.go

```diff
@@ -0,0 +1,105 @@
+// Copyright © 2022-2023 Obol Labs Inc. Licensed under the terms of a Business Source License 1.1
+
+package pidfile
+
+import (
+	"context"
+	"os"
+	"os/signal"
+	"path/filepath"
+	"sync/atomic"
+
+	"github.com/obolnetwork/charon/app/errors"
+	"github.com/obolnetwork/charon/app/log"
+	"github.com/obolnetwork/charon/app/z"
+)
+
+const (
+	filename = "charon-pidfile"
+)
+
+// New creates a pidfile called "charon-pidfile" in dataDir, writing contextStr in it.
+// If a pidfile exists already in dataDir New returns an error, a clean-up function otherwise.
+// New also registers a SIGINT signal handler so that it cleans its state if CTRL-C is called.
+func New(dataDir, contextStr string) (func() error, error) {
+	pfPath := filepath.Join(dataDir, filename)
+
+	if _, err := os.Stat(pfPath); err == nil {
+		readCtxStr, err := os.ReadFile(pfPath)
+		if err != nil {
+			return nil, errors.Wrap(err, "could not read pidfile content even if present")
+		}
+
+		return nil, errors.New(
+			"another instance of charon is running on the selected data directory",
+			z.Str("data_directory", dataDir),
+			z.Str("context", string(readCtxStr)),
+		)
+	} else if errors.Is(err, os.ErrNotExist) {
+		return createPidfile(pfPath, contextStr)
+	} else {
+		return nil, errors.Wrap(err, "fatal error while handling pidfile", z.Str("path", pfPath))
+	}
+}
+
+func createPidfile(path, contextStr string) (func() error, error) {
+	f, err := os.Create(path)
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot create pidfile", z.Str("path", path))
+	}
+
+	amt, err := f.WriteString(contextStr)
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot write context string in pidfile", z.Str("path", path))
+	}
+
+	if amt != len(contextStr) {
+		return nil, errors.New(
+			"could not write entirety of context string in pidfile",
+			z.Str("path", path),
+			z.Int("expected", len(contextStr)),
+			z.Int("got", amt),
+		)
+	}
+
+	alreadyDeleted := atomic.Bool{}
+	alreadyDeleted.Store(false)
+
+	ctx := context.Background()
+
+	c := make(chan os.Signal, 1)
+	signal.Notify(c, os.Interrupt)
+	go func() {
+		<-c
+
+		defer os.Exit(0)
+
+		if alreadyDeleted.Load() {
+			log.Debug(ctx, "Pidfile already deleted, not deleting from ctrl-c handler")
+			return
+		}
+
+		alreadyDeleted.Store(true)
```

Comment:
the delete might fail in which case it wasn't deleted

Reply:
I'll remove the SIGINT subscribe but just for the sake of clarity: in case deleting the file was not possible at this stage, it means there's something seriously wrong with the system that runs charon.

It's the edge case we were talking about when we originally discussed this: if we're failing here, the user needs to be aware of how to delete the pidfile and kickstart charon again.

This still holds true, because the deletion closure could fail anyway: it's a documentation problem that we will have to solve.